src/persist-client/src/internal/compact.rs

```diff
@@ -184,82 +192,416 @@ impl Compactor {
+                &mut compaction_parts_count,
+                desc.clone(),
+            )
+            .await;
+        }
+
+        let compaction_parts = batch_parts.finish().await;
+        assert_eq!(compaction_parts.len(), compaction_parts_count);
+        Ok((compaction_parts, compaction_runs, total_updates))
+    }
+
+    /// Consolidates `updates`, and determines whether the updates should extend the
+    /// current run (if any).  A new run will be created if `updates` contains a key
+    /// that overlaps with the current or any previous run.
+    fn consolidate_run<T, D>(
+        updates: &mut Vec<((Vec<u8>, Vec<u8>), T, D)>,
+        compaction_runs: &mut Vec<usize>,
+        number_of_compacted_runs: usize,
+        greatest_kv: &mut Option<(Vec<u8>, Vec<u8>)>,
+    ) -> usize
+    where
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64 + Send + Sync,
+    {
+        consolidate_updates(updates);
+
+        match (&greatest_kv, updates.last()) {
+            // our updates contain a key that exists within the range of a run we've
+            // already created, we should start a new run, as this part is no longer
+            // contiguous with the previous run/part
+            (Some(greatest_kv_seen), Some(greatest_kv_in_batch))
+                if *greatest_kv_seen > greatest_kv_in_batch.0 =>
+            {
+                compaction_runs.push(number_of_compacted_runs);
             }
-            let parts = parts.finish().await;
-
-            Ok(CompactRes {
-                output: HollowBatch {
-                    desc: req.desc,
-                    parts,
-                    len,
-                },
-            })
+            (_, Some(greatest_kv_in_batch)) => *greatest_kv = Some(greatest_kv_in_batch.0.clone()),
+            (Some(_), None) | (None, None) => {}
         };
 
-        // Compaction is cpu intensive, so be polite and spawn it on the
-        // CPU heavy runtime.
-        let compact_span = debug_span!("compact::consolidate");
-        cpu_heavy_runtime
-            .spawn_named(
-                || "persist::compact::consolidate",
-                compact_blocking.instrument(compact_span),
-            )
-            .await?
+        updates.len()
+    }
+
+    /// Encodes `updates` into columnar format and writes its parts out to blob.
+    /// The caller is expected to chunk `updates` into batches no greater than
+    /// [crate::PersistConfig::blob_target_size]
+    async fn write_run<T, D>(
+        batch_parts: &mut BatchParts<T>,
+        updates: &mut Vec<((Vec<u8>, Vec<u8>), T, D)>,
+        compaction_parts_count: &mut usize,
+        desc: Description<T>,
+    ) where
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64 + Send + Sync,
+    {
+        if updates.is_empty() {
+            return;
+        }
+        *compaction_parts_count += 1;
+
+        let mut builder = ColumnarRecordsVecBuilder::new_with_len(KEY_VAL_DATA_MAX_LEN);
```

Comment:
can you tie this KEY_VAL_DATA_MAX_LEN to the bit in the rustdoc about "The caller is expected to chunk `updates`" (or just move it here) plz?